source/1_engine/rex_engine/src/platform/win/filesystem/win_vfs.cpp

```diff
@@ -0,0 +1,637 @@
+          rsl::unique_ptr<QueuedRequest> request = rsl::move(g_read_requests[filepath]);
+          g_read_requests.erase(filepath);
+
+          // we don't need access to the queue anymore, we can unlock its access mutex
+          lock.unlock();
+
+          // read the actual file we requested
+          memory::Blob buffer = read_file(request->filepath());
+
+          // signal all read requests that this file has now been read
+          // it's possible multiple read requests want to access the same file
+          // if such requests come in while there's already a request for this file
+          // they get added to the original queued request and they now all get notified
+          request->signal_requests(buffer.data(), buffer.size());
+
+          // add the closed task to the queue, making sure it stays alive until all requests have finished processing the data
+          // This moves ownership of the buffer to the new request in the new queue
+          // We do this so that we don't have lock any thread while waiting for the requests to complete
+          const rsl::unique_lock closed_req_lock(g_closed_request_mutex);
+          g_closed_requests.push_back(rsl::move(request));
+        }
+
+        using namespace rsl::chrono_literals; // NOLINT(google-build-using-namespace)
+        rsl::this_thread::sleep_for(1ms);
+      }
+    }
+
+    void wait_for_read_requests()
+    {
+      while (g_vfs_state_controller.has_state(VfsState::Running))
+      {
+        const rsl::unique_lock lock(g_closed_request_mutex);
+        if (!g_closed_requests.empty())
+        {
+          // Go over all the queued request and check if their read requests are all done
+          for (rsl::unique_ptr<QueuedRequest>& request : g_closed_requests)
+          {
+            if (request->all_requests_finished())
+            {
+              request.reset();
+            }
+          }
+
+          // Remove all requests that are nullptr, aka all requests that have finished
+          auto it = rsl::remove(g_closed_requests.begin(), g_closed_requests.end(), nullptr);
+          g_closed_requests.erase(it, g_closed_requests.end());
+        }
+
+        using namespace rsl::chrono_literals; // NOLINT(google-build-using-namespace)
+        rsl::this_thread::sleep_for(1ms);
+      }
+    }
+
+    void start_threads()
+    {
+      // This thread processes the read requests
+      // A queued request is popped from the queue
+      // reads the file specified as the requested file to read
+      // signals all the read requests that this file is read
+      // and passed in the buffer with its size
+      // After which it waits async for all read requests to have finished processing
+      // the data passed in to them
+      // Finally it passes itself over to the closing thread queue
+      // to make sure the data stays alive until all requests have finished
+      g_reading_thread = rsl::thread(process_read_requests);
```

Comment:
INFO: This is probably already planned for the future, but wouldn't a job system be a better pick here? just spawning a new "job" that handles these read request. Would also make it possible to read multiple files at once.